File: backend/internal/auth/masterkey.go
// masterkey.go — the per-user MASTER KEY primitive (Tier-1 honest-privacy foundation).
//
// ─── What this is ─────────────────────────────────────────────────────────────
//
// Every account owns one random 256-bit MASTER KEY. It is the single root from
// which all future per-content encryption keys are derived (see DeriveContentKey
// and KEY HIERARCHY below). The master key itself is NEVER stored in the clear
// and NEVER logged; only two *wrapped* copies of it are persisted:
//
//	MASTER KEY ─┬─ AES-256-GCM wrap under a PBKDF2(password)-derived key   → env.Password
//	            └─ AES-256-GCM wrap under a HKDF(recovery-phrase)-derived key → env.Phrase
//
// Either wrap can reconstruct the same master key:
//   - normal unlock → UnwrapWithPassword (client-side at login, see src/lib/masterKey.js)
//   - password lost → UnwrapWithMnemonic (recovery-phrase path)
//   - password reset → RewrapPassword (unwrap via phrase, re-wrap the SAME master
//     key under the new password; the phrase wrap and therefore the master key
//     itself are unchanged, so previously-encrypted content stays decryptable).
//
// ─── KEY HIERARCHY (wave 3 builds on this) ────────────────────────────────────
//
//	recovery phrase (24-word BIP39)                password
//	        │ PBKDF2-HMAC-SHA512 (BIP39 seed)          │ PBKDF2-HMAC-SHA256
//	        │ + HKDF-SHA256                            │  (600k iters)
//	        ▼                                          ▼
//	  phraseWrapKey (32B)                        passwordWrapKey (32B)
//	        │  AES-256-GCM                              │  AES-256-GCM
//	        └──────────────► MASTER KEY (32B) ◄─────────┘
//	                              │ HKDF-SHA256, info="vulos-content:<domain>:<id>"
//	                              ▼
//	                        per-content key (32B)   ← WAVE 3 wraps content keys here
//
// ─── Crypto choices / WebCrypto parity ────────────────────────────────────────
//
// The password wrap uses PBKDF2-HMAC-SHA256 (not Argon2id) and AES-256-GCM (not
// XChaCha20-Poly1305) DELIBERATELY: both are natively available in the browser's
// WebCrypto SubtleCrypto with ZERO third-party JavaScript, which is what makes an
// honest, dependency-free CLIENT-SIDE unwrap possible (src/lib/masterKey.js). The
// envelope carries an explicit `kdf` tag and version so migrating the password
// KDF to Argon2id later (once an argon2 WASM module is bundled for the browser) is
// a forward-compatible change, not a breaking one. See SECURITY note in the PR.
//
// The recovery-phrase side REUSES the existing recovery.go BIP39 machinery
// (entropyToMnemonic / deriveRecoverySeed / ValidateMnemonic), so the master-key
// phrase is a real 24-word BIP39 phrase with the same properties as the rest of
// the recovery subsystem.
//
// Fail-closed: every unwrap authenticates via AES-GCM; a wrong password or wrong
// phrase fails the AEAD tag check and returns an error — it never returns partial
// or unauthenticated key material.
package auth

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"crypto/subtle"
	"encoding/binary"
	"encoding/json"
	"errors"
	"fmt"
	"io"

	"golang.org/x/crypto/hkdf"
)

// ─── Parameters (part of the wire format; do not change without a version bump) ─

const (
	// MasterKeyLen is the size of the per-user master key in bytes (256 bits).
	MasterKeyLen = 32

	// mkPBKDF2Iters is the PBKDF2-HMAC-SHA256 iteration count for the password
	// wrap. 600k follows OWASP 2023 guidance for PBKDF2-HMAC-SHA256.
	mkPBKDF2Iters = 600_000

	// masterKeyEnvelopeVersion is the on-disk envelope schema version.
	masterKeyEnvelopeVersion = 1

	kdfPassword = "pbkdf2-sha256"
	kdfPhrase   = "bip39-hkdf-sha256"
)

// AEAD associated-data domain separators. Distinct tags for the two wraps mean a
// blob sealed for one slot cannot be silently substituted into the other.
var (
	aadPasswordWrap = []byte("vulos-mk-pw-v1")
	aadPhraseWrap   = []byte("vulos-mk-phrase-v1")
	// hkdfPhraseInfo derives the phrase wrap key from the BIP39 seed.
	hkdfPhraseInfo = []byte("vulos-masterkey-wrap-v1")
	// hkdfZeroSalt is an explicit zero salt so Go and WebCrypto HKDF agree byte
	// for byte (WebCrypto requires an explicit salt parameter).
	hkdfZeroSalt = make([]byte, sha256.Size)
)

// ─── Envelope wire format ─────────────────────────────────────────────────────

// wrapSlot is one AES-256-GCM-sealed copy of the master key.
type wrapSlot struct {
	KDF  string `json:"kdf"`
	Iter int    `json:"iter,omitempty"` // PBKDF2 iterations (password slot only)
	Salt []byte `json:"salt,omitempty"` // PBKDF2 salt (password slot only)
	IV   []byte `json:"iv"`             // 12-byte AES-GCM nonce
	CT   []byte `json:"ct"`             // ciphertext || 16-byte tag
}

// MasterKeyEnvelope is the opaque, server-storable object. It never contains the
// master key or the recovery phrase — only two wrapped copies of the key.
type MasterKeyEnvelope struct {
	Version  int      `json:"v"`
	Password wrapSlot `json:"pw"`
	Phrase   wrapSlot `json:"phrase"`
}

// Marshal serialises the envelope to JSON bytes for persistence.
func (e *MasterKeyEnvelope) Marshal() ([]byte, error) { return json.Marshal(e) }

// ParseMasterKeyEnvelope parses a stored envelope blob.
func ParseMasterKeyEnvelope(blob []byte) (*MasterKeyEnvelope, error) {
	var e MasterKeyEnvelope
	if err := json.Unmarshal(blob, &e); err != nil {
		return nil, fmt.Errorf("auth/masterkey: parse envelope: %w", err)
	}
	if e.Version != masterKeyEnvelopeVersion {
		return nil, fmt.Errorf("auth/masterkey: unsupported envelope version %d", e.Version)
	}
	return &e, nil
}

// PasswordEnvelopeJSON returns just the password wrap slot as a compact JSON
// object suitable for handing to the browser so it can unwrap the master key
// client-side. It intentionally omits the phrase slot (that path is recovery-only
// and never needed for a normal login).
func (e *MasterKeyEnvelope) PasswordEnvelopeJSON() ([]byte, error) {
	return json.Marshal(map[string]any{
		"v":    e.Version,
		"kdf":  e.Password.KDF,
		"iter": e.Password.Iter,
		"salt": e.Password.Salt,
		"iv":   e.Password.IV,
		"ct":   e.Password.CT,
	})
}

// ─── Generation + wrapping ────────────────────────────────────────────────────

// NewMasterKey returns a fresh random 256-bit master key.
func NewMasterKey() ([]byte, error) {
	mk := make([]byte, MasterKeyLen)
	if _, err := io.ReadFull(rand.Reader, mk); err != nil {
		return nil, fmt.Errorf("auth/masterkey: generate: %w", err)
	}
	return mk, nil
}

// WrapMasterKey seals mk under BOTH the password and the recovery mnemonic and
// returns the persistable envelope. It does not retain mk.
func WrapMasterKey(mk []byte, password, mnemonic string) (*MasterKeyEnvelope, error) {
	if len(mk) != MasterKeyLen {
		return nil, fmt.Errorf("auth/masterkey: master key must be %d bytes", MasterKeyLen)
	}
	if password == "" {
		return nil, errors.New("auth/masterkey: password must not be empty")
	}
	if err := ValidateMnemonic(mnemonic); err != nil {
		return nil, fmt.Errorf("auth/masterkey: invalid mnemonic: %w", err)
	}

	pw, err := sealUnderPassword(mk, password)
	if err != nil {
		return nil, err
	}
	ph, err := sealUnderMnemonic(mk, mnemonic)
	if err != nil {
		return nil, err
	}
	return &MasterKeyEnvelope{Version: masterKeyEnvelopeVersion, Password: pw, Phrase: ph}, nil
}

// ProvisionMasterKey is the one-shot signup helper: it generates a fresh master
// key AND a fresh 24-word recovery phrase, wraps the key under both the password
// and that phrase, and returns the persistable envelope blob plus the phrase for
// ONE-TIME display to the user. The plaintext master key is zeroed before return;
// the phrase is returned to the caller (which must show it once and never persist
// it). Neither the master key nor the phrase is ever logged.
func ProvisionMasterKey(password string) (envelopeBlob []byte, mnemonic string, err error) {
	mk, err := NewMasterKey()
	if err != nil {
		return nil, "", err
	}
	defer zero(mk)

	entropy := make([]byte, 32)
	if _, err = io.ReadFull(rand.Reader, entropy); err != nil {
		return nil, "", fmt.Errorf("auth/masterkey: entropy: %w", err)
	}
	mnemonic, err = entropyToMnemonic(entropy)
	if err != nil {
		return nil, "", fmt.Errorf("auth/masterkey: mnemonic: %w", err)
	}

	env, err := WrapMasterKey(mk, password, mnemonic)
	if err != nil {
		return nil, "", err
	}
	blob, err := env.Marshal()
	if err != nil {
		return nil, "", fmt.Errorf("auth/masterkey: marshal: %w", err)
	}
	return blob, mnemonic, nil
}

// ─── Unwrapping ───────────────────────────────────────────────────────────────

// UnwrapWithPassword reconstructs the master key from the password wrap. This is
// the same operation the browser performs client-side at login; the Go version
// exists for server-side flows (reset) and for tests. A wrong password fails the
// AEAD tag check and returns an error (fail-closed).
func UnwrapWithPassword(e *MasterKeyEnvelope, password string) ([]byte, error) {
	if e == nil {
		return nil, errors.New("auth/masterkey: nil envelope")
	}
	if e.Password.KDF != kdfPassword {
		return nil, fmt.Errorf("auth/masterkey: unsupported password kdf %q", e.Password.KDF)
	}
	iter := e.Password.Iter
	if iter <= 0 {
		return nil, errors.New("auth/masterkey: invalid pbkdf2 iterations")
	}
	key := pbkdf2SHA256([]byte(password), e.Password.Salt, iter, 32)
	defer zero(key)
	return aesGCMOpen(key, e.Password.IV, e.Password.CT, aadPasswordWrap)
}

// UnwrapWithMnemonic reconstructs the master key from the recovery-phrase wrap.
func UnwrapWithMnemonic(e *MasterKeyEnvelope, mnemonic string) ([]byte, error) {
	if e == nil {
		return nil, errors.New("auth/masterkey: nil envelope")
	}
	if e.Phrase.KDF != kdfPhrase {
		return nil, fmt.Errorf("auth/masterkey: unsupported phrase kdf %q", e.Phrase.KDF)
	}
	if err := ValidateMnemonic(mnemonic); err != nil {
		return nil, fmt.Errorf("auth/masterkey: invalid mnemonic: %w", err)
	}
	key, err := phraseWrapKey(mnemonic)
	if err != nil {
		return nil, err
	}
	defer zero(key)
	return aesGCMOpen(key, e.Phrase.IV, e.Phrase.CT, aadPhraseWrap)
}

// RewrapPassword performs a phrase-based password reset: it unwraps the master
// key with the recovery mnemonic and re-wraps THE SAME master key under the new
// password. The phrase wrap is preserved unchanged, so the master key — and every
// piece of content ever encrypted under it — remains recoverable. A wrong
// mnemonic fails closed and no re-wrap happens.
func RewrapPassword(e *MasterKeyEnvelope, mnemonic, newPassword string) (*MasterKeyEnvelope, error) {
	if newPassword == "" {
		return nil, errors.New("auth/masterkey: new password must not be empty")
	}
	mk, err := UnwrapWithMnemonic(e, mnemonic)
	if err != nil {
		return nil, err
	}
	defer zero(mk)

	pw, err := sealUnderPassword(mk, newPassword)
	if err != nil {
		return nil, err
	}
	// Preserve the existing phrase slot verbatim: the master key is unchanged.
	return &MasterKeyEnvelope{Version: masterKeyEnvelopeVersion, Password: pw, Phrase: e.Phrase}, nil
}

// ─── Content-key derivation (wave 3 entry point) ──────────────────────────────

// DeriveContentKey derives a 256-bit per-content key from the master key using
// HKDF-SHA256 with a domain/id-scoped info string. Wave 3 (content encryption)
// wraps individual mail/file/etc. content keys under keys derived here, so that
// each content domain and item gets an independent key while everything remains
// rooted in the single master key. Deterministic: same (master, domain, id) →
// same key. domain and id MUST NOT contain the ':' separator ambiguously; they
// are length-prefixed to make the derivation injective.
func DeriveContentKey(masterKey []byte, domain, id string) ([]byte, error) {
	if len(masterKey) != MasterKeyLen {
		return nil, fmt.Errorf("auth/masterkey: master key must be %d bytes", MasterKeyLen)
	}
	info := make([]byte, 0, len(domain)+len(id)+32)
	info = append(info, []byte("vulos-content:")...)
	info = appendLenPrefixed(info, []byte(domain))
	info = appendLenPrefixed(info, []byte(id))
	hk := hkdf.New(sha256.New, masterKey, hkdfZeroSalt, info)
	out := make([]byte, 32)
	if _, err := io.ReadFull(hk, out); err != nil {
		return nil, fmt.Errorf("auth/masterkey: derive content key: %w", err)
	}
	return out, nil
}

// ─── Internals ────────────────────────────────────────────────────────────────

func sealUnderPassword(mk []byte, password string) (wrapSlot, error) {
	salt := make([]byte, 16)
	if _, err := io.ReadFull(rand.Reader, salt); err != nil {
		return wrapSlot{}, err
	}
	key := pbkdf2SHA256([]byte(password), salt, mkPBKDF2Iters, 32)
	defer zero(key)
	iv, ct, err := aesGCMSeal(key, mk, aadPasswordWrap)
	if err != nil {
		return wrapSlot{}, err
	}
	return wrapSlot{KDF: kdfPassword, Iter: mkPBKDF2Iters, Salt: salt, IV: iv, CT: ct}, nil
}

func sealUnderMnemonic(mk []byte, mnemonic string) (wrapSlot, error) {
	key, err := phraseWrapKey(mnemonic)
	if err != nil {
		return wrapSlot{}, err
	}
	defer zero(key)
	iv, ct, err := aesGCMSeal(key, mk, aadPhraseWrap)
	if err != nil {
		return wrapSlot{}, err
	}
	return wrapSlot{KDF: kdfPhrase, IV: iv, CT: ct}, nil
}

// phraseWrapKey derives the 32-byte AES key that wraps the master key under the
// recovery phrase: BIP39 seed (reused from recovery.go) → HKDF-SHA256.
func phraseWrapKey(mnemonic string) ([]byte, error) {
	seed := deriveRecoverySeed(normaliseMnemonic(mnemonic)) // recovery.go
	hk := hkdf.New(sha256.New, seed, hkdfZeroSalt, hkdfPhraseInfo)
	key := make([]byte, 32)
	if _, err := io.ReadFull(hk, key); err != nil {
		return nil, fmt.Errorf("auth/masterkey: derive phrase key: %w", err)
	}
	return key, nil
}

// aesGCMSeal returns (iv, ciphertext||tag). AES-256-GCM, 12-byte nonce — chosen
// for byte-for-byte parity with WebCrypto's AES-GCM.
func aesGCMSeal(key, plaintext, aad []byte) (iv, ct []byte, err error) {
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, nil, err
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, nil, err
	}
	iv = make([]byte, gcm.NonceSize())
	if _, err = io.ReadFull(rand.Reader, iv); err != nil {
		return nil, nil, err
	}
	ct = gcm.Seal(nil, iv, plaintext, aad)
	return iv, ct, nil
}

func aesGCMOpen(key, iv, ct, aad []byte) ([]byte, error) {
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, err
	}
	if len(iv) != gcm.NonceSize() {
		return nil, errors.New("auth/masterkey: bad nonce size")
	}
	pt, err := gcm.Open(nil, iv, ct, aad)
	if err != nil {
		// Uniform error: never distinguish "wrong key" from "tampered" — both are
		// simply "cannot decrypt" (fail-closed).
		return nil, errors.New("auth/masterkey: unwrap failed (wrong password or phrase?)")
	}
	if len(pt) != MasterKeyLen {
		zero(pt)
		return nil, errors.New("auth/masterkey: unwrapped key has wrong length")
	}
	return pt, nil
}

// pbkdf2SHA256 is an inline PBKDF2-HMAC-SHA256 (mirrors the SHA-512 variant in
// recovery.go) to avoid adding a golang.org/x/crypto/pbkdf2 import.
func pbkdf2SHA256(password, salt []byte, iter, keyLen int) []byte {
	hashLen := sha256.Size
	blocks := (keyLen + hashLen - 1) / hashLen
	buf := make([]byte, 0, blocks*hashLen)
	for block := 1; block <= blocks; block++ {
		var blockInt [4]byte
		binary.BigEndian.PutUint32(blockInt[:], uint32(block))
		mac := hmac.New(sha256.New, password)
		mac.Write(salt)
		mac.Write(blockInt[:])
		u := mac.Sum(nil)
		xored := make([]byte, len(u))
		copy(xored, u)
		for i := 1; i < iter; i++ {
			mac = hmac.New(sha256.New, password)
			mac.Write(u)
			u = mac.Sum(nil)
			for j := range xored {
				xored[j] ^= u[j]
			}
		}
		buf = append(buf, xored...)
	}
	return buf[:keyLen]
}

func appendLenPrefixed(dst, b []byte) []byte {
	var n [4]byte
	binary.BigEndian.PutUint32(n[:], uint32(len(b)))
	dst = append(dst, n[:]...)
	return append(dst, b...)
}

// zero overwrites b so key material does not linger in memory longer than needed.
func zero(b []byte) {
	for i := range b {
		b[i] = 0
	}
}

// constantTimeEqual is a small helper for tests / callers comparing key material.
func constantTimeEqual(a, b []byte) bool {
	return subtle.ConstantTimeCompare(a, b) == 1
}

File: src/lib/masterKey.js
// masterKey.js — CLIENT-SIDE master-key unwrap (WAVE2-RECOVERY, Tier-1).
//
// The server stores only a doubly-wrapped MASTER KEY envelope and never holds the
// key in the clear. At login the browser fetches the password-wrapped slot and
// unwraps it HERE, entirely client-side, using the platform WebCrypto SubtleCrypto
// API with ZERO third-party JavaScript. The recovered key lives only in memory for
// the session (see holdMasterKey) and is never written to storage or logged.
//
// Wire format is produced by backend/internal/auth/masterkey.go and is byte-for-
// byte compatible:
//   - password slot: PBKDF2-HMAC-SHA256(password, salt, iter) -> AES-256-GCM key
//   - phrase slot:   BIP39 seed = PBKDF2-HMAC-SHA512(mnemonic, "mnemonic", 2048)
//                    -> HKDF-SHA256(seed, zeroSalt, info) -> AES-256-GCM key
//   - content keys:  HKDF-SHA256(masterKey, zeroSalt, "vulos-content:<domain>:<id>")
//
// Fail-closed: every unwrap authenticates via AES-GCM; a wrong password/phrase or
// any tampering throws — no partial or unauthenticated key material is returned.

const AAD_PW = 'vulos-mk-pw-v1'
const AAD_PHRASE = 'vulos-mk-phrase-v1'
const HKDF_PHRASE_INFO = 'vulos-masterkey-wrap-v1'
const ZERO_SALT_32 = new Uint8Array(32)

const te = new TextEncoder()

function subtle() {
  const s = globalThis.crypto && globalThis.crypto.subtle
  if (!s) throw new Error('masterKey: WebCrypto SubtleCrypto unavailable (requires a secure context)')
  return s
}

// b64ToBytes decodes standard base64 (as emitted by Go's encoding/json for []byte).
function b64ToBytes(b64) {
  if (b64 instanceof Uint8Array) return b64
  const bin = atob(String(b64))
  const out = new Uint8Array(bin.length)
  for (let i = 0; i < bin.length; i++) out[i] = bin.charCodeAt(i)
  return out
}

// bytesToB64 encodes to standard base64 — the shape Go's encoding/json expects for
// a []byte field, so a slot produced here round-trips into the Go wrapSlot verbatim.
function bytesToB64(u8) {
  let bin = ''
  for (let i = 0; i < u8.length; i++) bin += String.fromCharCode(u8[i])
  return btoa(bin)
}

// PBKDF2_ITERS mirrors backend mkPBKDF2Iters — the standard password-wrap cost.
// The server enforces this as a FLOOR (mkMinPBKDF2Iters), rejecting anything lower.
const PBKDF2_ITERS = 600000

// normaliseMnemonic mirrors auth/recovery.go: lowercase + collapse whitespace.
function normaliseMnemonic(m) {
  return String(m).trim().toLowerCase().split(/\s+/).join(' ')
}

// unwrapMasterKeyWithPassword unwraps the master key from the password slot.
// `slot` is the object returned by GET /api/auth/masterkey/envelope:
//   { v, kdf, iter, salt, iv, ct }  (salt/iv/ct base64)
// Returns a Uint8Array(32). Throws on wrong password / tampering.
export async function unwrapMasterKeyWithPassword(slot, password) {
  if (!slot || slot.kdf !== 'pbkdf2-sha256') {
    throw new Error('masterKey: unsupported password kdf')
  }
  const iter = slot.iter | 0
  if (iter <= 0) throw new Error('masterKey: invalid iterations')
  const baseKey = await subtle().importKey('raw', te.encode(password), 'PBKDF2', false, ['deriveKey'])
  const aesKey = await subtle().deriveKey(
    { name: 'PBKDF2', salt: b64ToBytes(slot.salt), iterations: iter, hash: 'SHA-256' },
    baseKey,
    { name: 'AES-GCM', length: 256 },
    false,
    ['decrypt'],
  )
  const mk = await subtle().decrypt(
    { name: 'AES-GCM', iv: b64ToBytes(slot.iv), additionalData: te.encode(AAD_PW) },
    aesKey,
    b64ToBytes(slot.ct),
  )
  return new Uint8Array(mk)
}

// unwrapMasterKeyWithPhrase unwraps the master key from the recovery-phrase slot
// ({ kdf:'bip39-hkdf-sha256', iv, ct }). This is the client-side recovery path.
export async function unwrapMasterKeyWithPhrase(slot, mnemonic) {
  if (!slot || slot.kdf !== 'bip39-hkdf-sha256') {
    throw new Error('masterKey: unsupported phrase kdf')
  }
  const norm = normaliseMnemonic(mnemonic)
  // BIP39 seed: PBKDF2-HMAC-SHA512(mnemonic, "mnemonic", 2048, 64 bytes).
  const seedKey = await subtle().importKey('raw', te.encode(norm), 'PBKDF2', false, ['deriveBits'])
  const seed = new Uint8Array(await subtle().deriveBits(
    { name: 'PBKDF2', salt: te.encode('mnemonic'), iterations: 2048, hash: 'SHA-512' },
    seedKey,
    512,
  ))
  const hkdfKey = await subtle().importKey('raw', seed, 'HKDF', false, ['deriveKey'])
  const aesKey = await subtle().deriveKey(
    { name: 'HKDF', hash: 'SHA-256', salt: ZERO_SALT_32, info: te.encode(HKDF_PHRASE_INFO) },
    hkdfKey,
    { name: 'AES-GCM', length: 256 },
    false,
    ['decrypt'],
  )
  const mk = await subtle().decrypt(
    { name: 'AES-GCM', iv: b64ToBytes(slot.iv), additionalData: te.encode(AAD_PHRASE) },
    aesKey,
    b64ToBytes(slot.ct),
  )
  return new Uint8Array(mk)
}

// wrapMasterKeyWithPassword seals `masterKey` under `password`, producing a slot
// byte-for-byte compatible with backend/internal/auth/masterkey.go's password wrap
// (PBKDF2-HMAC-SHA256 -> AES-256-GCM, AAD 'vulos-mk-pw-v1'). Returns
// { v, kdf, iter, salt, iv, ct } with salt/iv/ct base64-encoded (matching Go's JSON
// []byte encoding), ready to POST. Runs entirely CLIENT-SIDE: the master key never
// leaves the tab and is never sent to the server — only this wrapped slot is.
export async function wrapMasterKeyWithPassword(masterKey, password, iter = PBKDF2_ITERS) {
  if (!(masterKey instanceof Uint8Array) || masterKey.length !== 32) {
    throw new Error('masterKey: master key must be 32 bytes')
  }
  if (!password) throw new Error('masterKey: password must not be empty')
  const salt = globalThis.crypto.getRandomValues(new Uint8Array(16))
  const iv = globalThis.crypto.getRandomValues(new Uint8Array(12))
  const baseKey = await subtle().importKey('raw', te.encode(password), 'PBKDF2', false, ['deriveKey'])
  const aesKey = await subtle().deriveKey(
    { name: 'PBKDF2', salt, iterations: iter, hash: 'SHA-256' },
    baseKey,
    { name: 'AES-GCM', length: 256 },
    false,
    ['encrypt'],
  )
  const ct = new Uint8Array(await subtle().encrypt(
    { name: 'AES-GCM', iv, additionalData: te.encode(AAD_PW) },
    aesKey,
    masterKey,
  ))
  return {
    v: 1,
    kdf: 'pbkdf2-sha256',
    iter,
    salt: bytesToB64(salt),
    iv: bytesToB64(iv),
    ct: bytesToB64(ct),
  }
}

// resetPasswordWithActiveSession performs the Tier-2 (active-session / trusted-
// device) password reset: it re-wraps the IN-MEMORY master key under `newPassword`
// CLIENT-SIDE and posts the wrapped slot + new password to the session-authed
// endpoint. The server never sees the master key — zero-access is preserved.
//
// If this session does not hold the master key (legacy login, or the key was
// cleared), it throws an error tagged `code === 'NO_MASTER_KEY'` so the caller can
// fall back to the recovery phrase (Tier-1) instead of silently failing.
export async function resetPasswordWithActiveSession(newPassword, fetchImpl = fetch) {
  const mk = getMasterKey()
  if (!mk) {
    const e = new Error('This session does not hold your encryption key — use your recovery phrase instead.')
    e.code = 'NO_MASTER_KEY'
    throw e
  }
  const slot = await wrapMasterKeyWithPassword(mk, newPassword)
  const res = await fetchImpl('/api/auth/masterkey/reset-active', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ new_password: newPassword, password_slot: slot }),
  })
  if (res.status === 409) {
    const e = new Error('This session does not hold your encryption key — use your recovery phrase instead.')
    e.code = 'NO_MASTER_KEY'
    throw e
  }
  if (!res.ok) {
    let msg = `reset failed (${res.status})`
    try {
      const j = await res.json()
      if (j && j.error) msg = j.error
    } catch { /* non-JSON error body */ }
    throw new Error(msg)
  }
  return { ok: true }
}

// deriveContentKey mirrors internal/auth.DeriveContentKey — the wave-3 entry
// point. HKDF-SHA256(masterKey, zeroSalt, "vulos-content:"+len(domain)+domain+
// len(id)+id). Returns a Uint8Array(32).
export async function deriveContentKey(masterKey, domain, id) {
  if (!(masterKey instanceof Uint8Array) || masterKey.length !== 32) {
    throw new Error('masterKey: master key must be 32 bytes')
  }
  const dom = te.encode(domain)
  const idb = te.encode(id)
  const info = new Uint8Array(14 + 4 + dom.length + 4 + idb.length)
  let o = 0
  info.set(te.encode('vulos-content:'), o); o += 14
  const dv = new DataView(info.buffer)
  dv.setUint32(o, dom.length); o += 4
  info.set(dom, o); o += dom.length
  dv.setUint32(o, idb.length); o += 4
  info.set(idb, o); o += idb.length

  const hkdfKey = await subtle().importKey('raw', masterKey, 'HKDF', false, ['deriveBits'])
  const bits = await subtle().deriveBits(
    { name: 'HKDF', hash: 'SHA-256', salt: ZERO_SALT_32, info },
    hkdfKey,
    256,
  )
  return new Uint8Array(bits)
}

// ─── In-memory master-key holder ──────────────────────────────────────────────
// The unlocked master key is kept ONLY in this module-level variable for the life
// of the tab. It is deliberately never placed in localStorage/sessionStorage/
// IndexedDB/cookies and never logged. Cleared on lock/logout via clearMasterKey().

let _masterKey = null

export function holdMasterKey(mk) {
  if (!(mk instanceof Uint8Array) || mk.length !== 32) {
    throw new Error('masterKey: refusing to hold a non-32-byte key')
  }
  _masterKey = mk
}

export function getMasterKey() {
  return _masterKey
}

export function hasMasterKey() {
  return _masterKey !== null
}

export function clearMasterKey() {
  if (_masterKey) _masterKey.fill(0)
  _masterKey = null
}

// unlockMasterKeyForSession fetches the password-wrapped envelope for the logged-
// in user and unwraps it client-side, holding the key in memory. Returns true on
// success, false if the account has no master key (legacy) — callers should treat
// a thrown error (wrong password / tamper) as fail-closed.
export async function unlockMasterKeyForSession(password, fetchImpl = fetch) {
  const res = await fetchImpl('/api/auth/masterkey/envelope')
  if (res.status === 404) return false // legacy account without a master key
  if (!res.ok) throw new Error(`masterKey: envelope fetch failed (${res.status})`)
  const slot = await res.json()
  const mk = await unwrapMasterKeyWithPassword(slot, password)
  holdMasterKey(mk)
  return true
}
